pkg/migration/runner.go

```diff
@@ -367,22 +369,59 @@ func (r *Runner) attemptMySQLDDL(ctx context.Context) error {
 		r.usedInstantDDL = true // success
 		return nil
 	}
-	// Inplace DDL is feature gated because it blocks replicas.
-	// It's only safe to do in aurora GLOBAL because replicas do not
-	// use the binlog.
-	if r.migration.AttemptInplaceDDL {
+
+	if err := r.safeInplaceAlter(); err != nil {
+		return err
+	} else {
 		err = r.attemptInplaceDDL(ctx)
 		if err == nil {
 			r.usedInplaceDDL = true // success
 			return nil
 		}
 	}
+
 	// Failure is expected, since MySQL DDL only applies in limited scenarios
 	// Return the error, which will be ignored by the caller.
 	// Proceed with regular copy algorithm.
 	return err
 }
 
+func (r *Runner) safeInplaceAlter() error {
+	// INPLACE DDL is not generally safe for online use in MySQL 8.0,
+	// because ADD INDEX can block replicas. Other INPLACE operations
+	// *are* safe, so we parse the ALTER and attempt to use INPLACE
+	// algorithm for any statements that do not add an index.
+
+	// Adding an index is only online-safe to do in Aurora GLOBAL
+	// because replicas do not use the binlog.
+
+	// If the operator has specified that they want to attempt
+	// an inplace add index, we will attempt inplace regardless
+	// of the statement.
+	if r.migration.AttemptInplaceAddIndex {
+		return nil
+	}
```

Comment:
I think it's easier to test this function if you move this out, so the caller then says:

```
if isASafeStatementForInplace(stmt) || r.migration.AttemptInplace {
```

In terms of the signature of the function, I think it should just accept an alter statement and return a boolean:

```
func AlgorithmInplaceConsideredSafe(stmt string) bool {}
```

I think we can put this in the utils package, so that it simplifies the migration package which is kind of large. We can then write tests for it independently.

Reply:
Fixed